src/services/codefixes/importFixes.ts

```diff
@@ -0,0 +1,393 @@
+                // Check to see if there are already imports being made from this source in the current file
+                const existingDeclaration = forEach(sourceFile.imports, importModuleSpecifier => {
+                    const importSymbol = checker.getSymbolAtLocation(importModuleSpecifier);
+                    if (importSymbol === moduleSymbol) {
+                        return getImportDeclaration(importModuleSpecifier);
+                    }
+                });
+
+                if (existingDeclaration) {
+                    return getCodeActionForExistingImport(existingDeclaration);
+                }
+                else {
+                    return getCodeActionForNewImport();
+                }
+
+                function getImportDeclaration(moduleSpecifier: LiteralExpression) {
+                    let node: Node = moduleSpecifier;
+                    while (node) {
+                        if (node.kind === SyntaxKind.ImportDeclaration) {
+                            return <ImportDeclaration>node;
+                        }
+                        if (node.kind === SyntaxKind.ImportEqualsDeclaration) {
+                            return <ImportEqualsDeclaration>node;
+                        }
+                        node = node.parent;
+                    }
+                    return undefined;
+                }
+
+                function getCodeActionForExistingImport(declaration: ImportDeclaration | ImportEqualsDeclaration): CodeAction {
+                    if (declaration.kind === SyntaxKind.ImportDeclaration) {
+                        const namedBindings = declaration.importClause && declaration.importClause.namedBindings;
+                        if (namedBindings) {
+                            if (namedBindings.kind === SyntaxKind.NamespaceImport) {
+                                return getCodeActionForNamespaceImport(namedBindings.name.getText());
+                            }
+                            /**
+                             * If the existing import declaration already has a named import list, just
+                             * insert the identifier into that list.
+                             */
+                            const textChange = getTextChangeForImportList(namedBindings);
+                            return createCodeAction(
+                                Diagnostics.Add_0_to_existing_import_declaration_from_1,
+                                [name, declaration.moduleSpecifier.getText()],
+                                textChange.newText,
+                                textChange.span,
+                                sourceFile.fileName
+                            );
+                        }
+
+                        // insert a new import statement in a new lines
+                        return getCodeActionForNewImport(declaration.moduleSpecifier.getText(), declaration.getEnd());

```

Comment:
Yes it is used when we want to add a new import list, while the existing one is a namespace import.

And in the latest implementation, when you have:
```ts
import * as ns from "module";
foo // foo is a member from ns
```
We provide two options:
1. change your code: so `foo` becomes `ns.foo`
2. add import list, so it becomes:
```ts
import * as ns from "module";
import { foo } from "module";
foo // foo is a member from ns
```
The user can choose from these two.

It is valid code, and as we cannot delete any user's code, I think these are the only options.